### ISSUE TITLE
TEP-0114: Custom Tasks Beta - Testing

### ISSUE THREAD

**Issue by jerop** [tektoncd/pipeline#5156]
In [TEP-0114: Custom Tasks Beta][tep-0114], we proposed migrating `Custom Tasks` and `Runs` to beta. 

Ahead of the promotion, we agreed to improve the test coverage for `Custom Tasks` and `Runs`. This issue tracks the testing work. 

Related issues: 
- https://github.com/tektoncd/pipeline/issues/5120
- https://github.com/tektoncd/pipeline/issues/4686

[tep-0114]: https://github.com/tektoncd/community/blob/main/teps/0114-custom-tasks-beta.md

**Comment by XinruZhang** [comment]
Currently, we have two ways to run custom tasks: w/ CRD, w/o CRD
- With CRD:  [task-loop](https://github.com/tektoncd/experimental/tree/main/task-loops)
- Without CRD: [wait-task](https://github.com/tektoncd/experimental/tree/main/wait-task)

The highest priority of this issue is to create an e2e test to ensure the basic functionalities of a custom task work as expected, which is a blocker of V1 Pipeline release. 

Then we can write a new CRD and controller with more functionalities implemented – as @lbernick suggested in issue #5120.

**Comment by XinruZhang** [comment]
- [ ] move wait-task to test/
- [ ] write e2e test (install tekton, custom controller)
- [ ] put this e2e test together with other tests? (not sure if this is required)
  - as part of a release test etc.
- [ ] documentation and future work (stage 2)

**Comment by XinruZhang** [comment]
After talking to @dibyom offline, we reached agreement on implementing a new controller for the testing purpose instead of using the exisitng `wait-task` because
1. `wait-task` will be eventually replaced by the new controller, no need to write code and make people review something that will definitely be removed;
3. the effort of implementing a new test controller is at the similar level of implementing a new controller.

**Comment by XinruZhang** [comment]
- [ ] Create a new controller under test/ with functionalities mentioned in #5120
- [ ] Write e2e test (install the controller, test cases)

**Comment by XinruZhang** [comment]
Next steps of improving the test coverage for Custom Task:
- [ ] Test Custom Task (without CRD) in PipelineRun
- [ ] Test Custom Task (with CRD) in TaskRun
- [ ] Test Custom Task (with CRD) in PipelineRun

**Comment by XinruZhang** [comment]
@jerop For the `retries`, we need to be aligned with the behavior of TaskRun, IMHO it would be better to implement it after addressing https://github.com/tektoncd/pipeline/issues/5248. wdyt?

**Comment by lbernick** [comment]
> For the `retries`, we need to be aligned with the behavior of TaskRun, IMHO it would be better to implement it after addressing #5248. wdyt?

I think since we've decided to add retries to TaskRun rather than changing how Run retries work, we can implement testing for Run retries in parallel. (We might want to wait for the TEP to be accepted, but I don't think we need to wait for the implementation to land.)

**Comment by XinruZhang** [comment]
@jerop I think we can mark this issue as done, wdyt?

**Comment by jerop** [comment]
yes, making it as done - thank you so much @XinruZhang 😻 

/close
